exchange/exchange.go

```diff
@@ -305,6 +312,125 @@ func (e *exchange) buildBidResponse(ctx context.Context, liveAdapters []openrtb_
+	if err != nil || primaryAdServer == "" {
+		return res, err
+	}
+
+	publisher := brandCatExt.Publisher
+
+	seatBidsToRemove := make([]openrtb_ext.BidderName, 0)
+
+	for bidderName, seatBid := range *seatBids {
+		bidsToRemove := make([]int, 0)
+		for bidInd := range seatBid.bids {
+			bid := seatBid.bids[bidInd].bid
+			var duration int
+			var category string
+			var tempRespBidExt openrtb_ext.ExtBid
+			if len(bid.Ext) > 0 {
+				if err := json.Unmarshal(bid.Ext, &tempRespBidExt); err != nil {
+					return res, err
+				}
+
+				//handler for different extension formats
+				var adapterExt json.RawMessage
+				if tempRespBidExt.Bidder != nil {
+					adapterExt = tempRespBidExt.Bidder
+				} else {
+					adapterExt = bid.Ext
+				}
+
+				//unmarshal bid extension to get video duration
+				var objmap map[string]*json.RawMessage
+				if err := json.Unmarshal(adapterExt, &objmap); err != nil {
+					return res, err
+				}
+				var bidderExt openrtb_ext.BidderExt
+				extData := objmap[bidderName.String()]
+				if err := json.Unmarshal(*extData, &bidderExt); err != nil {
+					return res, err
+				}
+				duration = bidderExt.CreativeInfo.Video.Duration
+
+			}
+
+			bidIabCat := bid.Cat
+			if len(bidIabCat) != 1 {
+				//TODO: add metrics
+				//on receiving bids from adapters if no unique IAB category is returned  or if no ad server category is returned discard the bid
+				bidsToRemove = append(bidsToRemove, bidInd)
+				continue
+			} else {
+				//if unique IAB category is present then translate it to the adserver category based on mapping file
+				category, err = categoriesFetcher.FetchCategories(primaryAdServer, publisher, bidIabCat[0])
+				if err != nil || category == "" {
+					//TODO: add metrics
+					//if mapping required but no mapping file is found then discard the bid
+					bidsToRemove = append(bidsToRemove, bidInd)
+					continue
+				}
+			}
+			categoryDuration := fmt.Sprintf("%s_%ds", category, duration)
+			res[bid.ID] = categoryDuration
+		}
+
+		if len(bidsToRemove) > 0 {
+			if len(bidsToRemove) == len(seatBid.bids) {
+				//if all bids are invalid - remove entire seat bid
+				seatBidsToRemove = append(seatBidsToRemove, bidderName)
+			} else {
+				bids := seatBid.bids
+				for i := len(bidsToRemove) - 1; i >= 0; i-- {
+					remInd := bidsToRemove[i]
+					bids = append(bids[:remInd], bids[remInd+1:]...)
+				}
+				seatBid.bids = bids
+			}
+		}
+
+	}
+	if len(seatBidsToRemove) > 0 {
+		if len(seatBidsToRemove) == len(*seatBids) {
+			//delete all seat bids
+			*seatBids = nil
+		} else {
+			for _, seatBidInd := range seatBidsToRemove {
+				delete(*seatBids, seatBidInd)
+			}
+
+		}
+	}
+
+	return res, nil
+}
+
+func getPrimaryAdServer(adServerId int) (string, error) {
+	switch adServerId {
+	case 1:
+		return "freewheel", nil
+	case 2:
+		return "dfp", nil
+	default:
+		return "", fmt.Errorf("Primary ad server is incorrect")
```

Comment:
Probably should change this to "Primary ad server is not recognized", or perhaps "unknown". Incorrect doesn't seem to be the right word to use.

Reply:
Updated, also added what the bad value was when it's unrecognized to the error message.